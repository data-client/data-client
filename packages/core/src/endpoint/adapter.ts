import { Endpoint } from '@rest-hooks/endpoint';
import type { EndpointInstance } from '@rest-hooks/endpoint';
import type { FetchShape } from '@rest-hooks/core/endpoint/shapes';

type ShapeTypeToSideEffect<T extends 'read' | 'mutate' | 'delete' | undefined> =
  T extends 'read' | undefined ? undefined : true;

const SIDEEFFECT_TYPES: (string | undefined)[] = ['mutate', 'delete'];

export default function shapeToEndpoint<
  Shape extends Partial<FetchShape<any, any>>,
>(
  shape: Shape,
): Shape['fetch'] extends (...args: any) => Promise<any>
  ? EndpointInstance<
      Shape['fetch'],
      Shape['schema'],
      ShapeTypeToSideEffect<Shape['type']>
    > &
      Shape['options']
  : Shape['options'] & { key: Shape['getFetchKey'] } {
  // make this identity function for endpoints
  if ((shape as any).key) return shape as any;
  const sideEffect = SIDEEFFECT_TYPES.includes(shape.type);
  const options = {
    ...shape.options,
    key: shape.getFetchKey,
    schema: shape.schema,
    ...((sideEffect && { sideEffect }) as any),
  };
  if (Object.prototype.hasOwnProperty.call(shape, 'fetch'))
    return new Endpoint(shape.fetch as any, options);

  return options as any;
}
