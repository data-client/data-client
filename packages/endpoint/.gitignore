/lib
/dist
/legacy
/index.d.ts
/ts3.4
/ts4.0
/ts4.2
