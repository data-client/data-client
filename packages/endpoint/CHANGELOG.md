# @data-client/endpoint

## 0.2.6

### Patch Changes

- 318df89bf7: Add nonFilterArgumentKeys argument to Collection

  `nonFilterArgumentKeys` defines a test to determine which argument keys
  are not used for filtering the results. For instance, if your API uses
  'orderBy' to choose a sort - this argument would not influence which
  entities are included in the response.

  This allows customizing `createCollectionFilter` for the
  most common case

- 1fcaeb1a7b: Fix some versions of RN not working with Collections

  (Array.sort() does not exist in all versions)

## 0.2.5

### Patch Changes

- 5a16f86668: Allowing Collection class override for createCollectionFilter default
- 4317adb94c: Collections with arguments in different orders now correctly mean the same Collection

  This could sometimes result in different instances of a Collection having different values.

## 0.2.4

### Patch Changes

- c8c5575e5a: Ignore 'undefined' parameters in collection matching
- 4e9d34ebc1: Optimistic creates no longer need a 'fake pk'

  e.g.,

  ```ts
  controller.fetch(TodoResource.getList.push, {
    // id: randomId(), THIS IS NO LONGER NEEDED
    userId,
    title: e.currentTarget.value,
  });
  ```

  This is achieved by computing a random id when a pk cannot be
  computed. In development mode non-create endpoints will
  still throw when pk fails.

## 0.2.3

### Patch Changes

- 8a71700644: Add EndpointToFunction type
- 8a71700644: scheam.push/unshift type denormalize to singular item (for now)
- 8a71700644: fix: Support Collections with boolean parameters

## 0.2.2

### Patch Changes

- f4b625df5a: schema.Entity.key should always be enumerable
- 8af1b5a8ef: Detect unusable pk when pk is serialized
- 6f3b39b585: Only warn about name mangling when using SSR and only once per type

## 0.2.1

### Patch Changes

- 69ce1f8b6b: docs: Add some more JSDoc links
- 12c2596453: fix: Ensure Entit.key can be set in v8 browsers
- 6c8f0be900: enhance: Include class/function in error message regarding name mangling

## 0.2.0

### Minor Changes

- bf141cb5a5: Removed deprecated Endpoint.optimisticUpdate -> use Endpoint.getOptimisticResponse
- 011cc20732: Remove FetchShape compatibility.
  This removes support for the legacy hooks in 'rest-hooks' like useResource()

### Patch Changes

- 9788090c55: Fix Collection creates (like push, unshift) normalizing when no collections are in store
- 9788090c55: Fix Collections handling single item with Union schema
