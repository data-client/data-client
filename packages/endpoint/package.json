{
  "name": "@rest-hooks/endpoint",
  "version": "2.2.10",
  "description": "Declarative Network Interface Definitions",
  "sideEffects": false,
  "main": "dist/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build:lib": "cross-env NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "cross-env NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:js:node": "cross-env BROWSERSLIST_ENV=node12 rollup -c",
    "build:js:browser": "cross-env BROWSERSLIST_ENV=legacy rollup -c",
    "build:bundle": "run-p build:js:* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib dist legacy ts3.4 *.tsbuildinfo",
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4 && copyfiles --up 1 ./src-legacy-types/**/*.d.ts ./ts3.4/",
    "dev": "yarn run build:lib -w",
    "prepare": "yarn run build:lib",
    "prepublishOnly": "yarn run build:bundle && yarn run build:legacy:lib",
    "test": "cross-env NODE_ENV=test jest",
    "test:ci": "yarn test -- --ci",
    "test:watch": "yarn test -- --watch",
    "test:coverage": "yarn test -- --coverage"
  },
  "keywords": [
    "data",
    "cache",
    "networking",
    "fetch",
    "typescript",
    "data fetching",
    "data cache",
    "api",
    "normalized cache",
    "subject pattern",
    "declarative",
    "definition",
    "schema",
    "async",
    "ajax",
    "react"
  ],
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "license": "Apache-2.0",
  "homepage": "https://resthooks.io/docs/api/Endpoint",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/endpoint"
  },
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "@rest-hooks/normalizr": "^8.2.9"
  }
}
