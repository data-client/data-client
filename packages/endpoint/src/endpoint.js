import { CSP } from './CSP.js';

function runCompat(endpoint, options) {
  endpoint.type = endpoint.sideEffect ? 'mutate' : 'read';
  endpoint.options = { ...options };
  delete endpoint.options.key;
  delete endpoint.options.schema;
  delete endpoint.options.sideEffect;
  delete endpoint.options.fetch;
  delete endpoint.options.getFetchKey;
  delete endpoint.options.options;
  if (Object.keys(endpoint.options).length === 0) {
    delete endpoint.options;
  }
  if (endpoint.schema === undefined) endpoint.schema = null;
}

/**
 * Defines an async data source.
 * @see https://resthooks.io/docs/api/Endpoint
 */
export default class Endpoint extends Function {
  constructor(fetchFunction, options) {
    let self;
    if (CSP) {
      self = (...args) => self.fetch(...args);
      Object.setPrototypeOf(self, new.target.prototype);
    } else {
      super('return arguments.callee.fetch.apply(arguments.callee, arguments)');
      self = this;
    }
    /** The following is for compatibility with FetchShape */
    self.getFetchKey = params => self.key(params);

    if (fetchFunction) self.fetch = fetchFunction;

    /** Name propery block
     *
     * To make things callable, we force every instance to be constructed as a function
     * Because of this the name property will be autoset
     * To create a usable naming inheritance pattern, we use __name as a proxy.
     * Every instance then overrides the name property.
     *
     * For protocol specific extensions that wish to customize default naming
     * behavior, be sure to add your own `Object.defineProperty(self, 'name'`
     * in your constructor to override this one.
     */
    let autoName;
    if (
      !(options && 'name' in options) &&
      fetchFunction &&
      fetchFunction.name &&
      fetchFunction.name !== 'anonymous'
    ) {
      autoName = fetchFunction.name;
    }
    Object.defineProperty(self, 'name', {
      get() {
        if (
          /* istanbul ignore else */ process.env.NODE_ENV !== 'production' &&
          self.key === Endpoint.prototype.key &&
          !(autoName || this.__name)
        ) {
          console.error(
            'Endpoint: Autonaming failure.\n\nEndpoint initialized with anonymous function.\nPlease add `name` option or hoist the function definition. https://resthooks.io/rest/api/Endpoint#name',
          );
        }
        return autoName || this.__name;
      },
      set(v) {
        this.__name = v;
      },
    });
    /** End name property block */

    Object.assign(self, options);

    /** The following is for compatibility with FetchShape */
    runCompat(self, options);
    return self;
  }

  key(...args) {
    return `${this.name} ${JSON.stringify(args)}`;
  }

  testKey(key) {
    return key.startsWith(this.name);
  }

  bind(thisArg, ...args) {
    const fetchFunc = this.fetch;
    const keyFunc = this.key;
    return this.extend({
      fetch() {
        return fetchFunc.apply(thisArg ?? this, args);
      },
      key() {
        return keyFunc.apply(this, args);
      },
    });
  }

  extend(options) {
    // make a constructor/prototype based off this
    // extend from it and init with options sent
    class E extends this.constructor {}

    Object.assign(E.prototype, this);
    const instance = new E(options.fetch, options);

    /** The following is for compatibility with FetchShape */
    runCompat(instance, { ...this.options, ...options });

    return instance;
  }

  /* istanbul ignore next */
  static {
    /* istanbul ignore if */
    if (runCompat.name !== 'runCompat') {
      this.prototype.key = function (...args) {
        console.error('Rest Hooks Error: https://resthooks.io/errors/osid');
        return `${this.name} ${JSON.stringify(args)}`;
      };
    }
  }
}
export const ExtendableEndpoint = Endpoint;
