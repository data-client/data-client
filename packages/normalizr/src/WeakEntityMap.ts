import { isImmutable } from './schemas/ImmutableUtils.js';
import { Path } from './types.js';

/** Maps entity dependencies to a value (usually their denormalized form)
 *
 * Dependencies store `Path` to enable quick traversal using only `State`
 * If *any* members of the dependency get cleaned up, so does that key/value pair get removed.
 */
export default class WeakEntityMap<K extends object = object, V = any> {
  readonly next = new WeakMap<K, Link<K, V>>();
  nextPath: Path | undefined = undefined;

  get(entity: K, getEntity: GetEntity<K | symbol>) {
    let curLink = this.next.get(entity);
    if (!curLink) return EMPTY;
    while (curLink.nextPath) {
      const nextEntity = getEntity(curLink.nextPath);
      curLink = curLink.next.get(nextEntity as any);
      if (!curLink) return EMPTY;
    }
    // curLink exists, but has no path - so must have a value
    return [curLink.value, curLink.journey] as [V, Path[]];
  }

  set(dependencies: Dep<K>[], value: V) {
    if (dependencies.length < 1) throw new KeySize();
    let curLink: Link<K, V> = this as any;
    for (const { entity, path } of dependencies) {
      let nextLink = curLink.next.get(entity);
      if (!nextLink) {
        nextLink = new Link<K, V>();
        curLink.next.set(entity, nextLink);
      }
      curLink.nextPath = path;
      curLink = nextLink;
    }
    // in case there used to be more
    curLink.nextPath = undefined;
    curLink.value = value;
    // we could recompute this on get, but it would have a cost and we optimize for `get`
    curLink.journey = depToPaths(dependencies);
  }
}

const EMPTY = [undefined, undefined] as const;

export function getEntities<K extends object>(state: State<K>): GetEntity<K> {
  const entityIsImmutable = isImmutable(state);

  if (entityIsImmutable) {
    return ({ key, pk }: Path) => state.getIn([key, pk]);
  } else {
    return ({ key, pk }: Path) => state[key]?.[pk];
  }
}

export function depToPaths(dependencies: Dep[]) {
  const paths = dependencies.map(dep => dep.path);
  // the first item is the result input not an actual entity
  if (paths[0]?.pk === '') paths.shift();
  return paths;
}

export type GetEntity<K = object | symbol> = (lookup: Path) => K;

/** Link in a chain */
class Link<K extends object, V> {
  next = new WeakMap<K, Link<K, V>>();
  value: V | undefined = undefined;
  journey: Path[] = [];
  nextPath: Path | undefined = undefined;
}

class KeySize extends Error {
  message = 'Keys must include at least one member';
}

export interface Dep<K = object> {
  path: Path;
  entity: K;
}

type State<K extends object> =
  | Record<string, Record<string, K>>
  | { getIn(path: [string, string]): K };
