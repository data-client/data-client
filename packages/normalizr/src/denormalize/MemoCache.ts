import { getEntities } from './getEntities.js';
import GlobalCache from './globalCache.js';
import getUnvisit from './unvisit.js';
import buildQueryKey, { validateQueryKey } from '../buildQueryKey.js';
import type { EntityTable, NormalizedIndex, Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type {
  DenormalizeNullable,
  EntityCache,
  EntityPath,
  EndpointsCache,
  NormalizeNullable,
} from '../types.js';
import WeakDependencyMap, { Dep, GetDependency } from '../WeakDependencyMap.js';

//TODO: make immutable distinction occur when initilizing MemoCache

/** Singleton to store the memoization cache for denormalization methods */
export default class MemoCache {
  /** Cache for every entity based on its dependencies and its own input */
  protected entities: EntityCache = {};
  /** Caches the final denormalized form based on input, entities */
  protected endpoints: EndpointsCache = new WeakDependencyMap<EntityPath>();
  /** Caches the queryKey based on schema, args, and any used entities or indexes */
  protected queryKeys: Record<string, WeakDependencyMap<QueryPath>> = {};

  /** Compute denormalized form maintaining referential equality for same inputs */
  denormalize<S extends Schema>(
    input: unknown,
    schema: S | undefined,
    entities: any,
    args: readonly any[] = [],
  ): {
    data: DenormalizeNullable<S> | symbol;
    paths: EntityPath[];
  } {
    // we already vary based on input, so we don't need endpointKey? TODO: verify
    // if (!this.endpoints[endpointKey])
    //   this.endpoints[endpointKey] = new WeakDependencyMap<EntityPath>();

    // undefined means don't do anything
    if (schema === undefined) {
      return { data: input as any, paths: [] };
    }
    if (input === undefined) {
      return { data: undefined as any, paths: [] };
    }
    const getEntity = getEntities(entities);

    return getUnvisit(
      getEntity,
      new GlobalCache(getEntity, this.entities, this.endpoints),
      args,
    )(input, schema);
  }

  /** Compute denormalized form maintaining referential equality for same inputs */
  query<S extends Schema>(
    argsKey: string,
    schema: S,
    args: any[],
    entities:
      | Record<string, Record<string, object>>
      | {
          get(k: string): any;
          getIn(k: string[]): any;
        },
    indexes:
      | NormalizedIndex
      | {
          getIn(k: string[]): any;
        },
  ): {
    data: DenormalizeNullable<S> | undefined;
    paths: EntityPath[];
    isInvalid: boolean;
  } {
    const input = this.buildQueryKey(argsKey, schema, args, entities, indexes);

    if (!validateQueryKey(input)) {
      return { data: input as any, paths: [], isInvalid: true };
    }

    const { data, paths } = this.denormalize(input, schema, entities, args);
    return { data, paths, isInvalid: false } as any;
  }

  buildQueryKey<S extends Schema>(
    argsKey: string,
    schema: S,
    args: any[],
    entities:
      | Record<string, Record<string, object>>
      | {
          get(k: string): any;
          getIn(k: string[]): any;
        },
    indexes:
      | NormalizedIndex
      | {
          getIn(k: string[]): any;
        },
  ): NormalizeNullable<S> {
    // This is redundant for buildQueryKey checks, but that was is used for recursion so we still need the checks there
    // null is object so we need double check
    if (
      (typeof schema !== 'object' &&
        typeof (schema as any).queryKey !== 'function') ||
      !schema
    )
      return schema as any;

    if (!this.queryKeys[argsKey]) {
      this.queryKeys[argsKey] = new WeakDependencyMap<QueryPath>();
    }
    // cache lookup: argsKey -> schema -> ...touched indexes or entities
    const queryCache = this.queryKeys[argsKey];
    const lookupEntity = createLookupEntity(entities);
    const lookupIndex = createLookupIndex(indexes);
    // eslint-disable-next-line prefer-const
    let [value, paths] = queryCache.get(
      schema as any,
      createDepLookup(lookupEntity, lookupIndex),
    );

    // paths undefined is the only way to truly tell nothing was found (the value could have actually been undefined)
    if (!paths) {
      // first dep path is ignored
      // we start with schema object, then lookup any 'touched' members and their paths
      const dependencies: Dep<QueryPath>[] = [
        { path: [''], entity: schema as any },
      ];

      value = buildQueryKey(
        schema,
        args,
        trackLookup(lookupEntity, dependencies),
        trackLookup(lookupIndex, dependencies),
      );
      queryCache.set(dependencies, value);
    }
    return value;
  }
}

type IndexPath = [key: string, field: string];
type EntitySchemaPath = [key: string];
type QueryPath = IndexPath | EntitySchemaPath;

function createDepLookup(lookupEntity, lookupIndex): GetDependency<QueryPath> {
  return (args: QueryPath) => {
    return args.length === 1 ? lookupEntity(...args) : lookupIndex(...args);
  };
}

function trackLookup<D extends any[], FD extends D>(
  lookup: (...args: FD) => any,
  dependencies: Dep<D>[],
) {
  return ((...args: Parameters<typeof lookup>) => {
    const value = lookup(...args);
    dependencies.push({ path: args, entity: value });
    return value;
  }) as any;
}

export function createLookupEntity(
  entities:
    | EntityTable
    | {
        get(k: string): { toJS(): any } | undefined;
      },
) {
  const entityIsImmutable = isImmutable(entities);
  if (entityIsImmutable) {
    return (entityKey: string) => entities.get(entityKey)?.toJS?.();
  } else {
    return (entityKey: string): { readonly [pk: string]: any } | undefined =>
      entities[entityKey];
  }
}

export function createLookupIndex(
  indexes:
    | NormalizedIndex
    | {
        getIn(k: string[]): any;
      },
) {
  const entityIsImmutable = isImmutable(indexes);
  if (entityIsImmutable) {
    return (
      key: string,
      field: string,
    ): { readonly [indexKey: string]: string | undefined } =>
      indexes.getIn([key, field])?.toJS?.();
  } else {
    return (
      key: string,
      field: string,
    ): { readonly [indexKey: string]: string | undefined } => {
      if (indexes[key]) {
        return indexes[key][field];
      }
      return {};
    };
  }
}
