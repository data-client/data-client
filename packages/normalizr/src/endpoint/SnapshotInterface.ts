import type { ExpiryStatusInterface } from '../Expiry.js';
import type { EndpointInterface } from './EndpointInterface.js';
import type { ErrorTypes } from './ErrorTypes.js';

export interface SnapshotInterface {
  getResponse: <
    E extends Pick<EndpointInterface, 'key' | 'schema' | 'invalidIfStale'>,
    Args extends readonly [...Parameters<E['key']>],
  >(
    endpoint: E,
    ...args: Args
  ) => {
    data: any;
    expiryStatus: ExpiryStatusInterface;
    expiresAt: number;
  };

  getError: <
    E extends Pick<EndpointInterface, 'key'>,
    Args extends readonly [...Parameters<E['key']>],
  >(
    endpoint: E,
    ...args: Args
  ) => ErrorTypes | undefined;

  readonly fetchedAt: number;
}
