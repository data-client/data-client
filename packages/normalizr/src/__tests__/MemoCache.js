// eslint-env jest
import { Entity } from '@data-client/endpoint';

import MemoCache from '../denormalize/MemoCache';
import WeakEntityMap from '../WeakEntityMap';

class IDEntity extends Entity {
  id = '';
  pk() {
    return this.id;
  }
}

class Tacos extends IDEntity {
  type = '';
}

let dateSpy;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe('denormalize with global cache', () => {
  test('denormalizes suspends when symbol contains DELETED string', () => {
    const entities = {
      Tacos: {
        1: Symbol('ENTITY WAS DELETED'),
      },
    };
    expect(new MemoCache().denormalize('1', Tacos, entities).data).toEqual(
      expect.any(Symbol),
    );
  });
  test('maintains referential equality with same results', () => {
    const memo = new MemoCache();
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
        2: { id: '2', type: 'bar' },
      },
    };
    const result = ['1', '2'];
    const schema = [Tacos];
    const { data: first } = memo.denormalize(result, schema, entities);
    const { data: second } = memo.denormalize(result, schema, entities);
    expect(first).toBe(second);

    const { data: third } = memo.denormalize([...result], schema, entities);
    expect(first).not.toBe(third);
    expect(first).toEqual(third);

    const fourth = memo.denormalize(result, schema, {
      Tacos: { ...entities.Tacos, 2: { id: '2', type: 'bar' } },
    }).data;
    expect(first).not.toBe(fourth);
    expect(first).toEqual(fourth);
  });

  describe('nested entities', () => {
    class User extends IDEntity {}
    class Comment extends IDEntity {
      comment = '';
      static schema = {
        user: User,
      };
    }
    class Article extends IDEntity {
      title = '';
      body = '';
      static schema = {
        author: User,
        comments: [Comment],
      };
    }

    const entities = {
      Article: {
        123: {
          author: '8472',
          body: 'This article is great.',
          comments: ['comment-123-4738'],
          id: '123',
          title: 'A Great Article',
        },
      },
      Comment: {
        'comment-123-4738': {
          comment: 'I like it!',
          id: 'comment-123-4738',
          user: '10293',
        },
      },
      User: {
        10293: {
          id: '10293',
          name: 'Jane',
        },
        8472: {
          id: '8472',
          name: 'Paul',
        },
      },
    };

    test('maintains referential equality with nested entities', () => {
      const memo = new MemoCache();

      const result = { data: '123' };
      const schema = { data: Article };
      const first = memo.denormalize(result, schema, entities).data;
      const second = memo.denormalize(result, schema, entities).data;
      expect(first).toBe(second);
      const third = memo.denormalize('123', Article, entities).data;
      const fourth = memo.denormalize('123', Article, entities).data;
      expect(third).toBe(fourth);
    });

    test('maintains responds to entity updates for distinct top-level results', () => {
      const memo = new MemoCache();

      const result1 = { data: '123' };
      const result2 = { results: ['123'] };
      const first = memo.denormalize(result1, { data: Article }, entities).data;
      const second = memo.denormalize(
        result2,
        { results: [Article] },
        entities,
      ).data;
      expect(first.data).toBe(second.results[0]);
      const third = memo.denormalize('123', Article, entities).data;
      expect(third).toBe(first.data);

      // now change
      const nextState = {
        ...entities,
        Article: {
          123: {
            ...entities.Article[123],
            title: 'updated article',
            body: 'new body',
          },
        },
      };
      const firstChanged = memo.denormalize(
        result1,
        { data: Article },
        nextState,
      ).data;
      expect(firstChanged).not.toBe(first);
      const secondChanged = memo.denormalize(
        result2,
        { results: [Article] },
        nextState,
      ).data;
      expect(secondChanged).not.toBe(second);
      expect(firstChanged.data).toBe(secondChanged.results[0]);
    });

    test('handles multi-schema (summary entities)', () => {
      class ArticleSummary extends IDEntity {
        title = '';
        body = '';
        static schema = {
          comments: [Comment],
        };

        static key = 'Article';
      }
      const memo = new MemoCache();
      const resultCacheA = new WeakEntityMap();
      const resultCacheB = new WeakEntityMap();
      const resultCacheC = new WeakEntityMap();

      // we have different result values to represent different endpoint inputs
      const resultA = { data: '123' };
      const resultB = { data: '123' };
      const resultC = '123';
      const firstSchema = { data: ArticleSummary };
      const secondSchema = { data: Article };
      const first = memo.denormalize(resultA, firstSchema, entities).data;
      const second = memo.denormalize(resultB, secondSchema, entities).data;
      // show distinction between how they are denormalized
      expect(first.data.author).toMatchInlineSnapshot(`"8472"`);
      expect(second.data.author).toMatchInlineSnapshot(`
        User {
          "id": "8472",
          "name": "Paul",
        }
      `);
      expect(first.data).not.toBe(second.data);
      const firstWithoutChange = memo.denormalize(
        resultA,
        firstSchema,
        entities,
      ).data;
      expect(first).toBe(firstWithoutChange);

      const third = memo.denormalize(resultC, Article, entities).data;
      expect(third).toBe(second.data);

      // now change
      const nextState = {
        ...entities,
        Article: {
          123: {
            ...entities.Article[123],
            title: 'updated article',
            body: 'new body',
          },
        },
      };
      const firstChanged = memo.denormalize(
        resultA,
        firstSchema,
        nextState,
      ).data;
      expect(firstChanged).not.toBe(first);
      const secondChanged = memo.denormalize(
        resultB,
        secondSchema,
        nextState,
      ).data;
      expect(secondChanged).not.toBe(second);
      expect(firstChanged.data.author).toMatchInlineSnapshot(`"8472"`);
      expect(secondChanged.data.author).toMatchInlineSnapshot(`
        User {
          "id": "8472",
          "name": "Paul",
        }
      `);
    });

    test('entity equality changes', () => {
      const memo = new MemoCache();

      const result = { data: '123' };
      const { data: first } = memo.denormalize(
        result,
        { data: Article },
        entities,
      );
      const { data: second } = memo.denormalize(
        result,
        { data: Article },
        {
          ...entities,
          Article: {
            123: {
              author: '8472',
              body: 'This article is great.',
              comments: ['comment-123-4738'],
              id: '123',
              title: 'A Great Article',
            },
          },
        },
      );
      expect(first).not.toBe(second);
      expect(first.data.author).toBe(second.data.author);
      expect(first.data.comments[0]).toBe(second.data.comments[0]);
    });

    test('nested entity equality changes', () => {
      const memo = new MemoCache();

      const result = { data: '123' };
      const { data: first } = memo.denormalize(
        result,
        { data: Article },
        entities,
      );

      const { data: second } = memo.denormalize(
        result,
        { data: Article },
        {
          ...entities,
          Comment: {
            'comment-123-4738': {
              comment: 'Updated comment!',
              id: 'comment-123-4738',
              user: '10293',
            },
          },
        },
      );

      expect(first).not.toBe(second);
      expect(first.data.title).toBe(second.data.title);
      expect(first.data.author).toBe(second.data.author);
      expect(second.data.comments[0].comment).toEqual('Updated comment!');
      expect(first.data.comments[0]).not.toBe(second.data.comments[0]);
      expect(first.data.comments[0].user).toBe(second.data.comments[0].user);
    });

    test('nested entity becomes present in entity table', () => {
      const memo = new MemoCache();

      const result = { data: '123' };
      const emptyEntities = {
        ...entities,
        // no Users exist
        User: {},
      };
      const { data: first } = memo.denormalize(
        result,
        { data: Article },
        emptyEntities,
      );

      const { data: second } = memo.denormalize(
        result,
        { data: Article },
        emptyEntities,
      );

      const { data: third } = memo.denormalize(
        result,
        { data: Article },
        // now has users
        entities,
      );

      expect(first.data.title).toBe(third.data.title);
      expect(first.data.author).toBeUndefined();
      // maintain cache when nested value is undefined
      expect(first.data).toBe(second.data);
      expect(first).toBe(second);
      // update value when nested value becomes defined
      expect(third.data.author).toBeDefined();
      expect(third.data.author.name).toEqual(expect.any(String));
      expect(first).not.toBe(third);
    });
  });

  test('nested entity becomes present in entity table with numbers', () => {
    const memo = new MemoCache();

    class User extends IDEntity {}
    class Article extends IDEntity {
      title = '';
      static schema = {
        author: User,
      };
    }
    const result = { data: 123 };
    const entities = {
      Article: {
        123: {
          author: 8472,
          id: 123,
          title: 'A Great Article',
        },
      },
      User: {
        8472: {
          id: 8472,
          name: 'Paul',
        },
      },
    };
    const emptyEntities = {
      ...entities,
      // no Users exist
      User: {},
    };
    const { data: first } = memo.denormalize(
      result,
      { data: Article },
      emptyEntities,
    );

    const { data: second } = memo.denormalize(
      result,
      { data: Article },
      emptyEntities,
    );

    const { data: third } = memo.denormalize(
      result,
      { data: Article },
      // now has users
      entities,
    );

    expect(first.data.title).toBe(third.data.title);
    expect(first.data.author).toBeUndefined();
    // maintain cache when nested value is undefined
    expect(first.data).toBe(second.data);
    expect(first).toBe(second);
    // update value when nested value becomes defined
    expect(third.data.author).toBeDefined();
    expect(third.data.author.name).toEqual(expect.any(String));
    expect(first).not.toBe(third);
  });

  test('denormalizes plain object with no entities', () => {
    const memo = new MemoCache();

    const input = {
      firstThing: { five: 5, seven: 42 },
      secondThing: { cars: 'fifo' },
    };
    const schema = {
      firstThing: { five: 0, seven: 0 },
      secondThing: { cars: '' },
    };
    const { data: first } = memo.denormalize(input, schema, {});
    expect(first).toEqual(input);
    // should maintain referential equality
    const { data: second } = memo.denormalize(input, schema, {});
    expect(second).toBe(first);
  });

  test('passthrough for null schema and an object input', () => {
    const memo = new MemoCache();

    const input = {
      firstThing: { five: 5, seven: 42 },
      secondThing: { cars: 'never' },
    };
    const { data } = memo.denormalize(input, null, {});
    expect(data).toBe(input);
  });

  test('passthrough for null schema and an number input', () => {
    const memo = new MemoCache();

    const input = 5;
    const { data } = memo.denormalize(input, null, {});
    expect(data).toBe(input);
  });

  test('passthrough for undefined schema and an object input', () => {
    const memo = new MemoCache();

    const input = {
      firstThing: { five: 5, seven: 42 },
      secondThing: { cars: 'never' },
    };
    const { data } = memo.denormalize(input, undefined, {});
    expect(data).toBe(input);
  });

  describe('null inputs when expecting entities', () => {
    class User extends IDEntity {}
    class Comment extends IDEntity {
      comment = '';
      static schema = {
        user: User,
      };
    }
    class Article extends IDEntity {
      title = '';
      body = '';
      author = User.fromJS({});
      comments = [];
      static schema = {
        author: User,
        comments: [Comment],
      };
    }

    test('handles null at top level', () => {
      const memo = new MemoCache();

      const denorm = memo.denormalize(null, { data: Article }, {}).data;
      expect(denorm).toEqual(null);
    });

    test('handles undefined at top level', () => {
      const memo = new MemoCache();

      const denorm = memo.denormalize(undefined, { data: Article }, {}).data;
      expect(denorm).toEqual(undefined);
    });

    test('handles null in nested place', () => {
      const memo = new MemoCache();

      const input = {
        data: { id: '5', title: 'hehe', author: null, comments: [] },
      };
      const denorm = memo.denormalize(input, { data: Article }, {}).data;
      expect(denorm).toMatchInlineSnapshot(`
        {
          "data": Article {
            "author": null,
            "body": "",
            "comments": [],
            "id": "5",
            "title": "hehe",
          },
        }
      `);
    });
  });
});
