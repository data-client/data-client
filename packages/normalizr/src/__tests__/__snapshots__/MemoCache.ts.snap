// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoCache query (direct) works with indexes 1`] = `
Cat {
  "id": "1",
  "name": "Milo",
  "username": "m",
}
`;

exports[`MemoCache query (direct) works with pk 1`] = `
Cat {
  "id": "1",
  "name": "Milo",
  "username": "m",
}
`;

exports[`MemoCache query (immutable) works with indexes 1`] = `
Cat {
  "id": "1",
  "name": "Milo",
  "username": "m",
}
`;

exports[`MemoCache query (immutable) works with pk 1`] = `
Cat {
  "id": "1",
  "name": "Milo",
  "username": "m",
}
`;
